docs/src/extend/plugin-migration-flat-config.md

```diff
@@ -0,0 +1,293 @@
+Environments are no longer supported in flat config, and so we recommend transitioning your environments into exported configs. For example, suppose you export a `mocha` environment like this:
+
+```js
+// plugin name: eslint-plugin-example
+module.exports = {
+    environments: {
+        mocha: {
+            globals: {
+                it: true,
+                xit: true,
+                describe: true,
+                xdescribe: true
+            }
+        }
+    },
+    rules: {
+        rule1: {},
+        rule2: {};
+    }
+};
+```
+
+To migrate this environment into a config, you need to add a new key in the `plugin.configs` object that has a flat config object containing the same information, like this:
+
+```js
+const plugin = {
+    configs: {},
+    rules: {},
+    processors: {}
+};
+
+// assign configs here so we can reference `plugin`
+Object.assign(plugin.configs, {
+    mocha: {
+        languageOptions: {
+            globals: {
+                it: "writeable",
+                xit: "writeable",
+                describe: "writeable",
+                xdescribe: "writeable"
+            }
+        }
+    }
+})
+
+// for ESM
+export default plugin;
+
+// OR for CommonJS
+module.exports = plugin;
+```
+
+Your users can then use this exported config like this:
+
+```js
+import example from "eslint-plugin-example";
+
+export default [
+
+    // use the mocha globals
+    example.configs.mocha,
+
+    // and provide your own overrides
+    {
+        languageOptions: {
+            globals: {
+                it: "readonly"
+            }
+        }
+    }
+];
+```
+
+You should update your documentation so your plugin users know how to reference the exported configs.
+
+## Backwards Compatibility
+
+If your plugin needs to work with both the old and new configuration systems, then you'll need to:
+
+1. **Export a CommonJS entrypoint.** The old configuration system cannot load plugins that are published only in ESM format. If your source code is in ESM, then you'll need to use a bundler that can generate a CommonJS version and use the [`exports`](https://nodejs.org/api/packages.html#package-entry-points) key in your `package.json` file to ensure the CommonJS version can be found by Node.js.
+1. **Keep the `environments` key.** If your plugin exports custom environments, you should keep those as they are and also export the equivalent flat configs as described above. The `environments` key is ignored when ESLint is running in flat config mode.
+1. **Export both eslintrc and flat configs.** The `configs` key is only validated when a config is used, so you can provide both formats of configs in the `configs` key. We recommend that you append older format configs with `-legacy` to make it clear that these configs will not be supported in the future. For example, if your primary config is called `recommended` and is in flat config format, then you can also have a config named `recommended-legacy` that is the eslintrc config format.
```

Comment:
I guess delineating the config name with `-legacy` is probably the absolute best way to ensure that you can keep the backwards compat and cleanly define both configs without muddling the new vs old keys and such.